### PR TITLE
libraries: tickv: Fixup the broken region search

### PR DESCRIPTION
**Pull request by alistair23**
### Pull Request Overview

Fixup the broken region search as described at:
https://github.com/tock/tock/pull/3617

We already have an existing test that accesses multiple regions, so we
also need to update that test now that we are correctly searching
regions.

As part of this change we convert the region from an isize to a uszie as
well.

This PR also fixes clippy warnings

### Testing Strategy

Unit tests, after the following commits have been removed as they break the tests

```
* 330aaede1 lib: tickv: update continue function comment
* 799c66057 tickv: need to return length for append
* 117a0a8ef tickv: async: flip err/ok
* bb3063da1 tickv: use value length
* a5e3f9973 tickv: remove get_buffer functions
* ba669a7f2 also need to return value buf on success
* 5273a6cc8 tickv: return value buffer on append fail
```

### Documentation Updated

- [X] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [X] Ran `make prepush`.

### CODE REVIEW

**Comment by bradjc** [comment]
Say the key hashes to something which ends with 5 0 bits and there are 32 regions. That will put the key into region 0. What is the search order if it is not found in region 0? Is it:
1. Look in region 0.
2. Look in region 1.
3. stop

? I don't think that matches what spec says but that is what this code does.

Also this code just searches the same page over and over for the same key.

I fixed the unit tests in #3619.

File: libraries/tickv/src/tickv.rs [hunk 1]
```diff
@@ -215,26 +215,40 @@ impl<'a, C: FlashController<S>, const S: usize> TicKV<'a, C, S> {
...
+                // If the offset is negative, convert to positive and increment by 1
                 region_offset if region_offset < 0 => -region_offset + 1,
                 _ => unreachable!(),
             };
 
             // Make sure our new offset is valid
-            if new_offset as usize > ((self.flash_size / S) - 1) {
+            if (region as isize + new_offset) as usize > ((self.flash_size / S) - 1) {
                 too_big = true;
                 continue;
```
Comment (bradjc) [inline libraries/tickv/src/tickv.rs:248]: Is this not just `break`?

File: libraries/tickv/src/tickv.rs [hunk 2]
```diff
@@ -672,17 +684,16 @@ impl<'a, C: FlashController<S>, const S: usize> TicKV<'a, C, S> {
                     KeyState::ReadRegion(reg) => reg as isize,
```
Comment (bradjc) [inline libraries/tickv/src/tickv.rs:684]: This will just keep returning the same region number over and over.

**Comment by bradjc** [comment]
https://github.com/tock/tock/pull/3623 should provide a test hopefully.

**Comment by alistair23** [comment]
Fixed up a few more corner cases and included the test from https://github.com/tock/tock/pull/3623

File: libraries/tickv/src/tickv.rs [hunk 3]
```diff
@@ -661,28 +681,33 @@ impl<'a, C: FlashController<S>, const S: usize> TicKV<'a, C, S> {
                 State::None => region as isize + region_offset,
                 State::Init(state) => {
                     match state {
-                        InitState::GetKeyReadRegion(reg) => reg as isize,
+                        InitState::GetKeyReadRegion(reg) => {
+                            region_offset = reg as isize - region as isize;
```
Comment (bradjc) [inline libraries/tickv/src/tickv.rs:685]:
```suggestion
```

File: libraries/tickv/src/tickv.rs [hunk 4]
```diff
@@ -661,28 +681,33 @@ impl<'a, C: FlashController<S>, const S: usize> TicKV<'a, C, S> {
...
                         _ => {
                             // Get the data from that region
                             region as isize + region_offset
                         }
                     }
                 }
                 State::GetKey(key_state) => match key_state {
-                    KeyState::ReadRegion(reg) => reg as isize,
+                    KeyState::ReadRegion(reg) => {
+                        region_offset = reg as isize - region as isize;
```
Comment (bradjc) [inline libraries/tickv/src/tickv.rs:696]:
```suggestion
```

File: libraries/tickv/src/tickv.rs [hunk 5]
```diff
@@ -765,9 +790,10 @@ impl<'a, C: FlashController<S>, const S: usize> TicKV<'a, C, S> {
                     self.read_buffer.replace(Some(region_data));
 
                     if cont {
-                        match self.increment_region_offset(new_region) {
+                        match self.increment_region_offset(region, region_offset) {
```
Comment (bradjc) [inline libraries/tickv/src/tickv.rs:793]:
```suggestion
                        region_offset = new_region as isize - region as isize;
                        match self.increment_region_offset(region, region_offset) {
```

File: libraries/tickv/src/tickv.rs [hunk 6]
```diff
@@ -661,28 +681,33 @@ impl<'a, C: FlashController<S>, const S: usize> TicKV<'a, C, S> {
                 State::None => region as isize + region_offset,
                 State::Init(state) => {
                     match state {
-                        InitState::GetKeyReadRegion(reg) => reg as isize,
+                        InitState::GetKeyReadRegion(reg) => {
+                            region_offset = reg as isize - region as isize;
+                            reg as isize
```
Comment (bradjc) [inline libraries/tickv/src/tickv.rs:686]:
```suggestion
                            reg
```

File: libraries/tickv/src/tickv.rs [hunk 7]
```diff
@@ -661,28 +681,33 @@ impl<'a, C: FlashController<S>, const S: usize> TicKV<'a, C, S> {
...
                             // Get the data from that region
                             region as isize + region_offset
                         }
                     }
                 }
                 State::GetKey(key_state) => match key_state {
-                    KeyState::ReadRegion(reg) => reg as isize,
+                    KeyState::ReadRegion(reg) => {
+                        region_offset = reg as isize - region as isize;
+                        reg as isize
```
Comment (bradjc) [inline libraries/tickv/src/tickv.rs:697]:
```suggestion
                        reg
```

Comment (bradjc) [reply libraries/tickv/src/tickv.rs:697]: Region cannot be a negative number.

File: libraries/tickv/src/tickv.rs [hunk 8]
```diff
@@ -661,28 +681,33 @@ impl<'a, C: FlashController<S>, const S: usize> TicKV<'a, C, S> {
                 State::None => region as isize + region_offset,
```
Comment (bradjc) [inline libraries/tickv/src/tickv.rs:681]:
```suggestion
                State::None => (region as isize + region_offset) as usize,
```

File: libraries/tickv/src/tickv.rs [hunk 9]
```diff
@@ -661,28 +681,33 @@ impl<'a, C: FlashController<S>, const S: usize> TicKV<'a, C, S> {
...
                 State::Init(state) => {
                     match state {
-                        InitState::GetKeyReadRegion(reg) => reg as isize,
+                        InitState::GetKeyReadRegion(reg) => {
+                            region_offset = reg as isize - region as isize;
+                            reg as isize
+                        }
                         _ => {
                             // Get the data from that region
                             region as isize + region_offset
```
Comment (bradjc) [inline libraries/tickv/src/tickv.rs:690]:
```suggestion
                            (region as isize + region_offset) as usize
```

File: libraries/tickv/src/tickv.rs [hunk 10]
```diff
@@ -661,28 +681,33 @@ impl<'a, C: FlashController<S>, const S: usize> TicKV<'a, C, S> {
...
+                    KeyState::ReadRegion(reg) => {
+                        region_offset = reg as isize - region as isize;
+                        reg as isize
+                    }
                 },
                 _ => unreachable!(),
-            };
+            }
+            .try_into()
+            .unwrap();
```
Comment (bradjc) [inline libraries/tickv/src/tickv.rs:703]:
```suggestion
            };
```

**Comment by bradjc** [comment]
Needs #3619 and #3623

**Review by bradjc — DISMISSED** [review]

**Comment by bradjc** [comment]
Ok I rebased and updated this PR.

Fixes tests, removes garbage collect state from append fn, and corrects init to be async.

**Review by bradjc — APPROVED** [review]
